nixos/modules/virtualisation/digital-ocean-config.nix

```diff
@@ -0,0 +1,167 @@
+      description = "Whether to fetch ssh keys from Digital Ocean";
+    };
+    seedEntropy = mkOption {
+      type = bool;
+      default = true;
+      example = true;
+      description = "Whether to run the kernel RNG entropy seeding script from the Digital Ocean vendor data";
+    };
+  };
+  config =
+    let
+      cfg = config.virtualisation.digitalOcean;
+      hostName = config.networking.hostName;
+    in mkMerge [{
+      fileSystems."/" = {
+        device = "/dev/disk/by-label/nixos";
+        autoResize = true;
+        fsType = "ext4";
+      };
+      boot = {
+        growPartition = true;
+        kernelParams = [ "console=ttyS0" "panic=1" "boot.panic_on_fail" ];
+        initrd.kernelModules = [ "virtio_scsi" ];
+        kernelModules = [ "virtio_pci" "virtio_net" ];
+        loader = {
+          grub.device = "/dev/sda";
+          timeout = 0;
+          grub.configurationLimit = 0;
+        };
+      };
+      services.openssh = {
+        enable = true;
+        permitRootLogin = "prohibit-password";
+        passwordAuthentication = mkDefault false;
+      };
+      networking = {
+        firewall.allowedTCPPorts = [ 22 ];
+        hostName = mkDefault ""; # use Digital Ocean metadata server
+        useNetworkd = true;
+      };
+
+      /* Fetch the root password from the digital ocean metadata.
+       * There is no specific route for this, so we use jq to get
+       * it from the One Big JSON metadata blob */
+      systemd.services.digitalocean-set-root-password = {
+        enable = cfg.setRootPassword;
+        path = [ pkgs.curl pkgs.shadow pkgs.jq ];
+        description = "Set root password provided by Digitalocean";
+        wantedBy = [ "multi-user.target" ];
+        script = ''
+          set -e
+          ROOT_PASSWORD=$(curl --retry-connrefused http://169.254.169.254/metadata/v1.json | jq -r '.auth_key')
+          echo "root:$ROOT_PASSWORD" | chpasswd
+          '';
+        unitConfig = {
+          After = [ "network-online.target" ];
+          Wants = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Set the hostname from Digital Ocean, unless the user configured it in
+       * the NixOS configuration */
+      systemd.services.digitalocean-set-hostname = {
+        enable = hostName == "";
+        path = [ pkgs.curl pkgs.nettools ];
+        description = "Set hostname provided by Digitalocean";
+        wantedBy = [ "multi-user.target" ];
+        script = ''
+          set -e
+          DIGITALOCEAN_HOSTNAME=$(curl --retry-connrefused http://169.254.169.254/metadata/v1/hostname)
+          hostname $DIGITALOCEAN_HOSTNAME
+        '';
+        unitConfig = {
+          After =  [ "network-online.target" ];
+          Wants = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Fetch the ssh keys for root from Digital Ocean */
```

Comment:
Can you reuse `nixos/modules/virtualisation/{openstack-config.nix,ec2-data,amazon-init.nix}` here? There's already a lot of metadata server fetching and applying…

Reply:
No not really.  They all are very specific to their platforms and I'm not sure how I could get a lot of code reuse out of them

Reply:
The `openstack-init` from `openstack-config.nix` uses `ec2-metadata-fetcher.nix`, which creates a service fetching metadata, which is picked up by the service defined in `ec2-data.nix` (which is imported).

This looks pretty similar to what we're trying to do here (and what we should do in `brightbox-image.nix` too, btw.)

Reply:
I made more than one attempt to use openstack to fetch metadata, and eventually settled on just using DigitalOcean's documented metdata service.